kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala

```diff
@@ -64,44 +69,76 @@ class KyuubiSessionImpl(
+
+    engineInitOp = sessionManager.operationManager.newInitEngineOperation(this)
+    _initEngineOpHandle = engineInitOp.getHandle
+    runOperation(engineInitOp)
+
     // we should call super.open after kyuubi session is already opened
     super.open()
   }
 
-  private def openSession(host: String, port: Int): Unit = {
-    val passwd = Option(password).filter(_.nonEmpty).getOrElse("anonymous")
-    val loginTimeout = sessionConf.get(ENGINE_LOGIN_TIMEOUT).toInt
-    val requestTimeout = sessionConf.get(ENGINE_REQUEST_TIMEOUT).toInt
-    transport = PlainSASLHelper.getPlainTransport(
-      user, passwd, new TSocket(host, port, requestTimeout, loginTimeout))
-    if (!transport.isOpen) {
-      transport.open()
-      logSessionInfo(s"Connected to engine [$host:$port]")
+  private[kyuubi] def openEngineSession(): Unit = synchronized {
+    withZkClient(sessionConf) { zkClient =>
+      val (host, port) = engine.getOrCreate(zkClient)
+      val passwd = Option(password).filter(_.nonEmpty).getOrElse("anonymous")
+      val loginTimeout = sessionConf.get(ENGINE_LOGIN_TIMEOUT).toInt
+      val requestTimeout = sessionConf.get(ENGINE_REQUEST_TIMEOUT).toInt
+      transport = PlainSASLHelper.getPlainTransport(
+        user, passwd, new TSocket(host, port, requestTimeout, loginTimeout))
+      if (!transport.isOpen) {
+        transport.open()
+        logSessionInfo(s"Connected to engine [$host:$port]")
+      }
+      client = new KyuubiSyncThriftClient(new TBinaryProtocol(transport))
+      // use engine SessionHandle directly
+      client.openSession(protocol, user, passwd, normalizedConf)
+      sessionManager.operationManager.setConnection(handle, client)
+      sessionEvent.openedTime = System.currentTimeMillis()
+      sessionEvent.sessionId = handle.identifier.toString
+      sessionEvent.clientVersion = handle.protocol.getValue
+      EventLoggingService.onEvent(sessionEvent)
     }
-    client = new KyuubiSyncThriftClient(new TBinaryProtocol(transport))
-    // use engine SessionHandle directly
-    _handle = client.openSession(protocol, user, passwd, normalizedConf)
-    sessionManager.operationManager.setConnection(handle, client)
-    sessionEvent.openedTime = System.currentTimeMillis()
-    sessionEvent.sessionId = handle.identifier.toString
-    sessionEvent.clientVersion = handle.protocol.getValue
-    EventLoggingService.onEvent(sessionEvent)
   }
 
   override protected def runOperation(operation: Operation): OperationHandle = {
-    sessionEvent.totalOperations += 1
+    if (operation.getHandle != _initEngineOpHandle) {
+      waitForEngineInitOpFinished()
+      sessionEvent.totalOperations += 1
+    }
     super.runOperation(operation)
   }
 
+  private def waitForEngineInitOpFinished(): Unit = {
+    if (!engineInitFinished) {
+      Option(engineInitOp).foreach { op =>
+        while (!OperationState.isTerminal(op.getStatus.state)) {
+          info("The engine init operation has not finished")
```

Comment:
I think debug is enough and we can add a new info logging before and after `while`.

Reply:
and we can aslo record the elapsed time